fix(multi_object_tracker): make sure smallest unknown footprint is 0.3

mebasoglu opened the pull request:
## Description

Related issue:
- https://github.com/autowarefoundation/autoware.universe/issues/6617

When unknown object polygon has only one point which is 0,0,0 , the node dies.

## Related links

<!-- Write the links related to this PR. Private links should be clearly marked as private, for example, '[FOO COMPANY INTERNAL LINK](https://example.com)'. -->

Fixes: https://github.com/autowarefoundation/autoware.universe/issues/6617

## Tests performed

<!-- Describe how you have tested this PR. -->

Confirmed on logging simulator.

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

There are two places where this error occurs, on `data_association.cpp` and `multi_object_tracker_core.cpp`. The fix was added those.

## Interface changes

<!-- Describe any changed interfaces, such as topics, services, or parameters. -->

## Effects on system behavior

<!-- Describe how this PR affects the system behavior. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

yukkysaito commented on perception/multi_object_tracker/src/multi_object_tracker_core.cpp:
```diff
@@ -386,6 +386,15 @@ void MultiObjectTracker::sanitizeTracker(
 
+      if (
+        object1.shape.type == autoware_auto_perception_msgs::msg::Shape::POLYGON &&
+        tier4_autoware_utils::getArea(object1.shape) == 0 &&
```
It is not very desirable to make equal comparisons between floating-point types and direct numbers. 🙏 
In this case, it is 0, so it is OK, but it may cause unintended bugs.

yukkysaito replied:
reference : https://www.geeksforgeeks.org/problem-in-comparing-floating-point-numbers-and-how-to-compare-them-correctly/

xmfcx replied:
@yukkysaito can we use the following directly directly:

https://github.com/autowarefoundation/autoware.universe/blob/52b2fd6a2099f0ff2fc0e9b9c22a9a14163ad849/common/autoware_auto_common/include/autoware_auto_common/helper_functions/float_comparisons.hpp#L106-L122

Or do we have these float comparisons somewhere else?

yukkysaito commented:
Thanks for your PR 👍 
We'll check this

armaganarsln commented:
/ask rate this PR quality, from 1 to 10. explain why

armaganarsln commented:
/review

armaganarsln commented:
/improve --extended

armaganarsln commented:
/ask At line 391 of multi_object_tracker_core.cpp, I'm comparing a float against an integer `tier4_autoware_utils::getArea(object1.shape) == 0` I don't think this is a good thing. What should I do instead?

xmfcx commented on perception/multi_object_tracker/src/data_association/data_association.cpp:
```diff
@@ -207,6 +207,13 @@ Eigen::MatrixXd DataAssociation::calcScoreMatrix(
           const double min_union_iou_area = 1e-2;
+          if (
+            measurement_object.shape.type == autoware_auto_perception_msgs::msg::Shape::POLYGON &&
+            area == 0 && measurement_object.shape.footprint.points.size() == 1) {
```
```suggestion
            area == 0 && measurement_object.shape.footprint.points.size() == 1) {
```
Same as here https://github.com/autowarefoundation/autoware.universe/pull/6618/files#r1525002210

yukkysaito commented on perception/multi_object_tracker/src/data_association/data_association.cpp:
```diff
@@ -207,6 +207,13 @@ Eigen::MatrixXd DataAssociation::calcScoreMatrix(
         if (passed_gate) {
           const double min_iou = min_iou_matrix_(tracker_label, measurement_label);
           const double min_union_iou_area = 1e-2;
+          if (
+            measurement_object.shape.type == autoware_auto_perception_msgs::msg::Shape::POLYGON &&
+            area == 0 && measurement_object.shape.footprint.points.size() == 1) {
+            measurement_object.shape.footprint.points.front().x = 0.3;
+            measurement_object.shape.footprint.points.front().y = 0.3;
+            measurement_object.shape.footprint.points.front().z = 0.3;
+          }
```
To set it to 0.3 would embed an unintended bug.In fact.
It sometimes results in strange behavior when there are actually objects with smaller values.
I think a better way to handle this issue is to have the getIntersectionArea function return 0 if the polygon does not have 3 or more points in it.

YoshiRi reviewed (commented):
I think this issue should be solved with fixing IoU caluculate function, so I do not want to apply this PR.
See https://github.com/autowarefoundation/autoware.universe/issues/6617#issuecomment-2001244853 .

mebasoglu commented:
Hello everyone, thank you for the feedback. I have pushed a commit for `get2dIoU`. Now, if there aren't enough points for a polygon, it will return `0.0`. It works on logging simulator.

The other approach was checking points size in `getIntersectionArea` function. That will also affect `get2dGeneralizedIoU`, 
`get2dPrecision` and `get2dRecall`.

What do you think about applying the patch to `getIntersectionArea`, or should it be only for `get2dIoU`?

YoshiRi commented:
@mebasoglu 
Thanks for your work!

As you say, changing  `getIntersectionArea`  may better since most function you mentioned has following escape sequence. I prefer this approach.
```cpp
  const double intersection_area = getIntersectionArea(source_polygon, target_polygon);
  if (intersection_area == 0.0) return 0.0; // you can escape here
```

But it seems `get2dGeneralizedIoU` do not have the escape, so we need to do some work then.

```cpp
  const double intersection_area = getIntersectionArea(source_polygon, target_polygon); // no escape
  const double union_area = getUnionArea(source_polygon, target_polygon); // this part may cause error
  const double convex_shape_area = getConvexShapeArea(source_polygon, target_polygon); // this will cause error
```

It would be helpful if you could change `getUnionArea`, `getConvexShapeArea` and `getIntersectionArea`.

technolojin reviewed (changes requested):
I agree to add checkers insures that the two inputs are valid.

The boost documentation of [intersection](https://live.boost.org/doc/libs/1_56_0/libs/geometry/doc/html/geometry/reference/algorithms/intersection.html) mentioned as following.
"Check the [Polygon Concept](https://live.boost.org/doc/libs/1_56_0/libs/geometry/doc/html/geometry/reference/concepts/concept_polygon.html) for the rules that polygon input for this algorithm should fulfill"

Having at least 3 footprint points is a necessary condition, not the sufficient condition. to have a area, the sufficient condition is to have a rank of 2.

I would like to suggest to add checkers of the validity of the polygon, as following.
```
static const double MIN_AREA = 1e-9;
if (boost::geometry::area(source_polygon) < MIN_AREA) return 0.0;
if (boost::geometry::area(target_polygon) < MIN_AREA) return 0.0;
```
In the official [documentation](https://live.boost.org/doc/libs/1_50_0/libs/geometry/doc/html/geometry/reference/algorithms/area/area_2_with_strategy.html) of the area function, 
" 
[Behavior](https://live.boost.org/doc/libs/1_50_0/libs/geometry/doc/html/geometry/reference/algorithms/area/area_2_with_strategy.html#geometry.reference.algorithms.area.area_2_with_strategy.behavior)
pointlike (e.g. point) Returns 0
linear (e.g. linestring) Returns 0
areal (e.g. polygon) Returns the area
"

The checkers may implied in the necessary functions, such as `get2dIoU`, `get2dGeneralizedIoU`, `get2dPrecision`, and `get2dRecall`.

mebasoglu commented:
Hello @YoshiRi @technolojin , thank you very much for the feedback.

I have created a new function `validatePolygons` . It returns `false` if the polygon areas are very close to zero, `true` otherwise. I added checks with this function on `getConvexShapeArea`, `getIntersectionArea` and `getUnionArea`. They return `0.0` if polygons aren't validated.

The `get2dGeneralizedIoU` function was also updated for the checks.

Tested on logging simulator.

technolojin commented:
@mebasoglu Thank you for your update. 
Your solution is a simple solution and it is important. However, I would like to give you some feedback.

1) Numerical error
As @mebasoglu checked, comparing a float number against zero (even the type is float) have a risk since the target number (`intersection_area` for example) can have a numerical error (overflow if this number is used for division, and that is what we do here). 

   Here are some related discussions.
   https://softwareengineering.stackexchange.com/questions/309350/avoiding-division-by-zero-using-float-comparison
   https://en.wikipedia.org/wiki/Machine_epsilon

  > Question: at line 391 of multi_object_tracker_core.cpp, i'm comparing a float against an integer `tier4_autoware_utils::getarea(object1.shape) == 0` i don't think this is a good thing. what should i do instead?
  > 
  > Answer: Comparing a float to an integer directly can indeed lead to precision issues due to the way floating point numbers are represented in memory. A better approach would be to use a small threshold value for the comparison.
  > 
  > Instead of `tier4_autoware_utils::getArea(object1.shape) == 0`, you could use something like:
  > 
  > ```c++
  > const float epsilon = 1e-9; // or any other small value that suits your precision needs
  > if (std::abs(tier4_autoware_utils::getArea(object1.shape)) < epsilon) {
  >     // ...
  > }
  > ```
  > 
  > This way, you're checking if the area is "close enough" to zero, taking into account the precision of floating point numbers.
  > **✨ Ask tool usage guide:**


2) Meaning of returning `0.0` on `getConvexShapeArea` and `getUnionArea` functions
As your updated functions, `getConvexShapeArea` and `getUnionArea` returns zero even one polygon have a valid area. As future users, it is confusing result. You implemented to check if the obtained area is zero before calculate necessary ratios, but other future developer may not understand the meaning and can cause catastrophic error (divide-by-zero). 
(Returning zero on `getIntersectionArea` is valid, by its definition.)

3) Duplicated checks
In `get2dGeneralizedIoU`,  the `validatePolygons` runs three times. and it is not necessary. 
The IoU calculations will repeated many many times and this change can make overall object tracking heavy, especially the number of tracking and detected object is large. 
(For example, calculation of matching score matrix have complexity of O(n^2))

I would like to request an efficient solution. Even it becomes a bit complicate, I think it worth.

mebasoglu commented:
Closing due to https://github.com/autowarefoundation/autoware.universe/pull/6701 was merged.
